### ISSUE TITLE
Incorrect Loading of Saved Dataset

### ISSUE THREAD

**Issue by Deech08** [jkrogager/VoigtFit#37]
I'm having trouble saving/loading a dataset that is fit with several components on several ions. 

Here is the output of a completed fit:

  Fit is running... Please, be patient.

 The fit has finished with the following exit message:
  Fit succeeded.

>   Best fit parameters
> 
> 				b			log(N)
> AlII  1670
> v = -44.2 +/- 24.4		 16.37 +/- 109.43	11.526 +/- 1.950
> v =   3.5 +/- 1.4		 17.95 +/-   3.14	12.986 +/- 0.066
> v =  45.0 +/- 1.3		 20.40 +/-   5.30	12.832 +/- 0.093
> v =  88.4 +/- 0.0		 11.51 +/-  27.23	11.996 +/- 0.846
> v = 122.6 +/- 0.0		 35.36 +/- 607.38	11.296 +/- 5.458
> v = 173.5 +/- 0.0		 50.00 +/- 191.95	12.023 +/- 1.694
> v = 239.0 +/- 0.0		 16.20 +/-   7.88	12.719 +/- 0.335
> v = 269.8 +/- 0.0		 24.19 +/-   7.89	12.984 +/- 0.145
> v = 323.3 +/- 0.0		 17.52 +/-   6.32	12.439 +/- 0.123
> 
> CII  1334
> v = -37.2 +/- 25.9		 20.12 +/-  19.58	14.116 +/- 1.267
> v =  12.9 +/- 10.2		 33.98 +/-  50.86	14.788 +/- 0.501
> v =  55.9 +/- 7.7		 10.34 +/-  20.54	14.393 +/- 0.615
> v =  88.4 +/- 2.0		  9.07 +/-   9.40	14.037 +/- 0.182
> v = 122.6 +/- 5.9		 18.52 +/-  21.38	13.890 +/- 0.733
> v = 173.5 +/- 39.0		 38.40 +/- 212.80	13.615 +/- 2.356
> v = 239.0 +/- 2.8		 25.28 +/-  25.28	14.199 +/- 1.883
> v = 269.8 +/- 6.5		 41.19 +/-  23.11	14.759 +/- 2.711
> v = 323.3 +/- 2.1		 15.35 +/-   6.92	14.288 +/- 0.325
> v = 371.6 +/- 2.5		 24.27 +/-   4.43	13.901 +/- 0.075
> 
> CIIa  1335.7, 1335.71
> v =  12.9 +/- 10.2		 33.98 +/-  50.86	13.880 +/- 15.567
> 
> FeII  1144, 1608
> v =   3.5 +/- 1.4		 26.96 +/-   3.62	14.261 +/- 0.056
> v =  45.0 +/- 1.3		 44.71 +/- 298.96	12.847 +/- 1.646
> v = 239.0 +/- 0.0		 22.37 +/-   6.18	13.872 +/- 0.226
> v = 269.8 +/- 0.0		 39.96 +/-   4.78	14.187 +/- 0.114
> 
> OI  1302
> v = -44.2 +/- 24.4		 49.51 +/- 490.96	13.361 +/- 3.605
> v =   3.5 +/- 1.4		 33.90 +/-  17.15	15.036 +/- 0.161
> v =  45.0 +/- 1.3		  9.00 +/-   2.94	14.027 +/- 0.802
> v =  88.4 +/- 0.0		 35.49 +/-  81.52	13.669 +/- 0.790
> v = 239.0 +/- 0.0		 30.04 +/-  14.01	14.569 +/- 0.372
> v = 269.8 +/- 0.0		 30.74 +/-  18.65	14.881 +/- 0.228
> v = 323.3 +/- 0.0		 18.73 +/-  20.79	14.142 +/- 0.431
> v = 371.6 +/- 0.0		 38.95 +/- 114.80	13.756 +/- 0.726
> 
> SiII  1190, 1193, 1260, 1304
>       1526
> v = -44.2 +/- 24.4		 44.94 +/-  18.28	12.824 +/- 0.318
> v =   3.5 +/- 1.4		 24.89 +/-   1.61	14.307 +/- 0.031
> v =  45.0 +/- 1.3		 13.09 +/-   1.35	13.942 +/- 0.058
> v =  88.4 +/- 2.0		 12.77 +/-   4.22	12.753 +/- 0.101
> v = 122.6 +/- 5.9		 23.25 +/-  19.34	12.435 +/- 0.796
> v = 173.5 +/- 39.0		 49.99 +/-  58.44	12.647 +/- 0.564
> v = 239.0 +/- 2.8		 15.12 +/-   2.84	13.984 +/- 0.185
> v = 269.8 +/- 6.5		 27.45 +/-   6.38	14.142 +/- 0.133
> v = 323.3 +/- 2.1		 17.35 +/-   2.04	13.491 +/- 0.070
> v = 371.6 +/- 2.5		 19.49 +/-   3.62	12.673 +/- 0.069
> 
> SiIII  1206
> v = -44.2 +/- 24.4		 44.94 +/-  18.28	12.638 +/- 0.249
> v =   3.5 +/- 1.4		 24.89 +/-   1.61	13.171 +/- 0.108
> v =  45.0 +/- 1.3		 13.09 +/-   1.35	14.256 +/- 0.230
> v =  88.4 +/- 0.0		 12.77 +/-   4.22	13.125 +/- 0.108
> v = 122.6 +/- 0.0		 23.25 +/-  19.34	12.759 +/- 0.621
> v = 173.5 +/- 0.0		 49.99 +/-  58.44	12.802 +/- 0.588
> v = 239.0 +/- 0.0		 15.12 +/-   2.84	13.107 +/- 0.320
> v = 269.8 +/- 0.0		 27.45 +/-   6.38	13.495 +/- 0.158
> v = 323.3 +/- 0.0		 17.35 +/-   2.04	13.462 +/- 0.100
> v = 371.6 +/- 0.0		 19.49 +/-   3.62	12.997 +/- 0.058

After I complete the fitting process, I save the dataset to a file 

`dataset.save("test.hdf5")`

Later when I try to load this same data file using

`dataset = VoigtFit.load_dataset("test.hdf5")`

The best fit parameters are mixed up. All the numbers are accurate, but the match up of z, b, logN for some component are shuffled around (such as the second listed component for Al II).
Here is a print out of the best fit parameters

>   Best fit parameters
> 
> 				b			log(N)
> AlII  1670
> v = -44.2 +/- 24.4		 16.37 +/- 109.43	11.526 +/- 1.950
> v = 323.3 +/- 1.4		 17.95 +/-   3.14	12.986 +/- 0.066
> v =   3.5 +/- 1.3		 20.40 +/-   5.30	12.832 +/- 0.093
> v =  45.0 +/- 0.0		 11.51 +/-  27.23	11.996 +/- 0.846
> v =  55.9 +/- 0.0		 35.36 +/- 607.38	11.296 +/- 5.458
> v =  88.4 +/- 0.0		 50.00 +/- 191.95	12.023 +/- 1.694
> v = 122.6 +/- 0.0		 16.20 +/-   7.88	12.719 +/- 0.335
> v = 173.5 +/- 0.0		 24.19 +/-   7.89	12.984 +/- 0.145
> v = 239.0 +/- 0.0		 17.52 +/-   6.32	12.439 +/- 0.123
> 
> CII  1334
> v = -37.2 +/- 25.9		 20.12 +/-  19.58	14.116 +/- 1.267
> v = 371.6 +/- 2.5		 24.27 +/-   4.43	13.901 +/- 0.075
> v =  12.9 +/- 10.2		 33.98 +/-  50.86	14.788 +/- 0.501
> v =  55.9 +/- 7.7		 10.34 +/-  20.54	14.393 +/- 0.615
> v =  88.4 +/- 2.0		  9.07 +/-   9.40	14.037 +/- 0.182
> v = 122.6 +/- 5.9		 18.52 +/-  21.38	13.890 +/- 0.733
> v = 173.5 +/- 39.0		 38.40 +/- 212.80	13.615 +/- 2.356
> v = 239.0 +/- 2.8		 25.28 +/-  25.28	14.199 +/- 1.883
> v = 269.8 +/- 6.5		 41.19 +/-  23.11	14.759 +/- 2.711
> v = 323.3 +/- 2.1		 15.35 +/-   6.92	14.288 +/- 0.325
> 
> CIIa  1335.7, 1335.71
> v = 371.6 +/- 10.2		 24.27 +/-  50.86	13.880 +/- 15.567
> 
> FeII  1144, 1608
> v = 323.3 +/- 1.4		 26.96 +/-   3.62	14.261 +/- 0.056
> v =   3.5 +/- 1.3		 44.71 +/- 298.96	12.847 +/- 1.646
> v = 122.6 +/- 0.0		 22.37 +/-   6.18	13.872 +/- 0.226
> v = 173.5 +/- 0.0		 39.96 +/-   4.78	14.187 +/- 0.114
> 
> OI  1302
> v = -44.2 +/- 24.4		 49.51 +/- 490.96	13.361 +/- 3.605
> v = 323.3 +/- 1.4		 33.90 +/-  17.15	15.036 +/- 0.161
> v =   3.5 +/- 1.3		  9.00 +/-   2.94	14.027 +/- 0.802
> v =  45.0 +/- 0.0		 35.49 +/-  81.52	13.669 +/- 0.790
> v = 122.6 +/- 0.0		 30.04 +/-  14.01	14.569 +/- 0.372
> v = 173.5 +/- 0.0		 30.74 +/-  18.65	14.881 +/- 0.228
> v = 239.0 +/- 0.0		 18.73 +/-  20.79	14.142 +/- 0.431
> v = 269.8 +/- 0.0		 38.95 +/- 114.80	13.756 +/- 0.726
> 
> SiII  1190, 1193, 1260, 1304
>       1526
> v = -44.2 +/- 24.4		 44.94 +/-  18.28	12.824 +/- 0.318
> v = 323.3 +/- 2.5		 19.49 +/-   3.62	12.673 +/- 0.069
> v =   3.5 +/- 1.4		 24.89 +/-   1.61	14.307 +/- 0.031
> v =  45.0 +/- 1.3		 13.09 +/-   1.35	13.942 +/- 0.058
> v =  55.9 +/- 2.0		 12.77 +/-   4.22	12.753 +/- 0.101
> v =  88.4 +/- 5.9		 23.25 +/-  19.34	12.435 +/- 0.796
> v = 122.6 +/- 39.0		 49.99 +/-  58.44	12.647 +/- 0.564
> v = 173.5 +/- 2.8		 15.12 +/-   2.84	13.984 +/- 0.185
> v = 239.0 +/- 6.5		 27.45 +/-   6.38	14.142 +/- 0.133
> v = 269.8 +/- 2.1		 17.35 +/-   2.04	13.491 +/- 0.070
> 
> SiIII  1206
> v = -44.2 +/- 24.4		 44.94 +/-  18.28	12.638 +/- 0.249
> v = 269.8 +/- 0.0		 17.35 +/-   3.62	12.997 +/- 0.058
> v = 323.3 +/- 1.4		 19.49 +/-   1.61	13.171 +/- 0.108
> v =   3.5 +/- 1.3		 24.89 +/-   1.35	14.256 +/- 0.230
> v =  45.0 +/- 0.0		 13.09 +/-   4.22	13.125 +/- 0.108
> v =  55.9 +/- 0.0		 12.77 +/-  19.34	12.759 +/- 0.621
> v =  88.4 +/- 0.0		 23.25 +/-  58.44	12.802 +/- 0.588
> v = 122.6 +/- 0.0		 49.99 +/-   2.84	13.107 +/- 0.320
> v = 173.5 +/- 0.0		 15.12 +/-   6.38	13.495 +/- 0.158
> v = 239.0 +/- 0.0		 27.45 +/-   2.04	13.462 +/- 0.100
> 

This same process works perfectly fine for ~20 other instances, but having this problem with 2 instances (both with several components to fit)

**Comment by jkrogager** [comment]
Hey, thanks for getting in touch. Just a few follow-up questions: 
What version of Python are you using?
Do you use any linking of parameters?
Are any parameters fixed during the fitting?

(On a side note: it seems that you may be over-fitting given the huge uncertainties for some components, consider using a lower number of components)

**Comment by Deech08** [comment]
Thanks for the quick follow up!

I'm using python 3.8.5 with voigtfit v3.11.7.1

Many parameters have links, but none are fixed. I think the easiest way for me to show this are just the lines that add the velocity components to voigtfit, which are below. 

I am additionally applying two additional constraints to keep the b values between 9 and 50 km/s and v between -150 and 450 km/s (converted to a relative redshift), which i'm doing directly to the dataset.pars with lmfit, before starting the fit process

And yes i agree on the over-fitting - this direction looks like a nightmare as well and i don't trust most of the components below 150 km/s, but my science case is only for the high velocity ones, which are the ones I prioritized more.

Thanks!

```
data.dataset.add_component_velocity("CII", -30., 20., 14.)
data.dataset.add_component_velocity("CII", -3., 20., 14.5)
data.dataset.add_component_velocity("CII", 50., 30., 14.5)
data.dataset.add_component_velocity("CII", 90., 30., 13.7)
data.dataset.add_component_velocity("CII", 120., 16., 13.5)
data.dataset.add_component_velocity("CII", 160., 19., 13.5)
data.dataset.add_component_velocity("CII", 225., 16., 14.)
data.dataset.add_component_velocity("CII", 280., 25., 14.)
data.dataset.add_component_velocity("CII", 330., 15., 14.)
data.dataset.add_component_velocity("CII", 385., 19., 13.75)

data.dataset.add_component_velocity("CIIa", -3, 20, 14., tie_z = "z1_CII", tie_b = "b1_CII")


data.dataset.add_component_velocity("SiII", -30., 20., 13.5)
data.dataset.add_component_velocity("SiII", -3., 20., 14.4)
data.dataset.add_component_velocity("SiII", 50., 20., 14.0)
data.dataset.add_component_velocity("SiII", 90., 20., 12.9, tie_z = "z3_CII")
data.dataset.add_component_velocity("SiII", 120., 16., 12.8, tie_z = "z4_CII")
data.dataset.add_component_velocity("SiII", 160., 19., 12.9, tie_z = "z5_CII")
data.dataset.add_component_velocity("SiII", 225., 16., 14.3, tie_z = "z6_CII")
data.dataset.add_component_velocity("SiII", 280., 25., 14.2, tie_z = "z7_CII")
data.dataset.add_component_velocity("SiII", 330., 15., 13.8, tie_z = "z8_CII")
data.dataset.add_component_velocity("SiII", 385., 19., 12.8, tie_z = "z9_CII")

data.dataset.add_component_velocity("SiIII", -30., 20., 12.5, tie_z = "z0_SiII", tie_b = "b0_SiII")
data.dataset.add_component_velocity("SiIII", -3., 20., 13.3, tie_z = "z1_SiII", tie_b = "b1_SiII")
data.dataset.add_component_velocity("SiIII", 50., 20., 13.4, tie_z = "z2_SiII", tie_b = "b2_SiII")
data.dataset.add_component_velocity("SiIII", 90., 20., 13.0, tie_z = "z3_SiII", tie_b = "b3_SiII")
data.dataset.add_component_velocity("SiIII", 120., 16., 12.8, tie_z = "z4_SiII", tie_b = "b4_SiII")
data.dataset.add_component_velocity("SiIII", 160., 19., 12.6, tie_z = "z5_SiII", tie_b = "b5_SiII")
data.dataset.add_component_velocity("SiIII", 225., 16., 13.3, tie_z = "z6_SiII", tie_b = "b6_SiII")
data.dataset.add_component_velocity("SiIII", 280., 25., 13.3, tie_z = "z7_SiII", tie_b = "b7_SiII")
data.dataset.add_component_velocity("SiIII", 330., 15., 13.3, tie_z = "z8_SiII", tie_b = "b8_SiII")
data.dataset.add_component_velocity("SiIII", 385., 19., 12.9, tie_z = "z9_SiII", tie_b = "b9_SiII")


data.dataset.add_component_velocity("FeII", -3., 20., 14.2, tie_z = "z1_SiII")
data.dataset.add_component_velocity("FeII", 50., 16., 13.4, tie_z = "z2_SiII")
data.dataset.add_component_velocity("FeII", 225., 19., 14.1, tie_z = "z6_SiII")
data.dataset.add_component_velocity("FeII", 280., 16., 13.9, tie_z = "z7_SiII")


data.dataset.add_component_velocity("AlII", -30., 20., 12., tie_z = "z0_SiII")
data.dataset.add_component_velocity("AlII", -3., 20., 13.0, tie_z = "z1_SiII")
data.dataset.add_component_velocity("AlII", 50., 20., 12.7, tie_z = "z2_SiII")
data.dataset.add_component_velocity("AlII", 90., 20., 12.0, tie_z = "z3_SiII")
data.dataset.add_component_velocity("AlII", 120., 16., 11.8, tie_z = "z4_SiII")
data.dataset.add_component_velocity("AlII", 160., 19., 11.8, tie_z = "z5_SiII")
data.dataset.add_component_velocity("AlII", 225., 16., 12.9, tie_z = "z6_SiII")
data.dataset.add_component_velocity("AlII", 280., 25., 12.8, tie_z = "z7_SiII")
data.dataset.add_component_velocity("AlII", 330., 15., 12.5, tie_z = "z8_SiII")

data.dataset.add_component_velocity("OI", -30., 20., 14.1, tie_z = "z0_SiII")
data.dataset.add_component_velocity("OI", -3., 20., 15., tie_z = "z1_SiII")
data.dataset.add_component_velocity("OI", 50., 20., 14.3, tie_z = "z2_SiII")
data.dataset.add_component_velocity("OI", 90., 20., 13.6, tie_z = "z3_SiII")
data.dataset.add_component_velocity("OI", 225., 16., 14.7, tie_z = "z6_SiII")
data.dataset.add_component_velocity("OI", 280., 25., 14.7, tie_z = "z7_SiII")
data.dataset.add_component_velocity("OI", 330., 15., 14.4, tie_z = "z8_SiII")
data.dataset.add_component_velocity("OI", 385., 15., 13.7, tie_z = "z9_SiII")
```

**Comment by jkrogager** [comment]
Thanks for the additional info. I'll have a look to see if I can reproduce this behaviour and find a fix.
I have a big deadline today though, so I might not have time until tomorrow. Sorry about that.

**Comment by Deech08** [comment]
Thanks so much and no problem with a wait at all!

**Comment by jkrogager** [comment]
The bug should be fixed now (latest version 3.11.8). Sorry for the delay.
